plugins/filter_aws/aws.c

```diff
@@ -183,190 +221,266 @@ static int cb_aws_init(struct flb_filter_instance *f_ins,
     return 0;
 }
 
-/* Get an IMDSv2 token */
-static int get_ec2_token(struct flb_filter_aws *ctx)
+/* Get VPC metadata, it called IMDS twice.
+ * First is for getting the Mac ID and combine into the path for VPC.
+ * Second call is using the VPC path to get the VPC id
+ */
+static int get_vpc_metadata(struct flb_filter_aws *ctx)
 {
     int ret;
-    size_t b_sent;
-    struct flb_connection *u_conn;
-    struct flb_http_client *client;
+    flb_sds_t mac_id = NULL;
+    flb_sds_t vpc_path;
+    flb_sds_t vpc_path_mem;
 
-    u_conn = flb_upstream_conn_get(ctx->ec2_upstream);
-    if (!u_conn) {
-        flb_plg_error(ctx->ins, "connection initialization error");
+    mac_id = flb_aws_imds_get_vpc_id(ctx->client_imds);
+    if (!mac_id) {
+        flb_plg_error(ctx->ins, "Could not retrieve vpc id from IMDS");
         return -1;
     }
 
-    /* Compose HTTP Client request */
-    client = flb_http_client(u_conn, FLB_HTTP_PUT,
-                             FLB_FILTER_AWS_IMDS_V2_TOKEN_PATH,
-                             NULL, 0, FLB_FILTER_AWS_IMDS_HOST,
-                             80, NULL, 0);
-
-    if (!client) {
-        flb_plg_error(ctx->ins, "count not create http client");
-        flb_upstream_conn_release(u_conn);
+    /* the VPC full path should be like:
+     *latest/meta-data/network/interfaces/macs/{mac_id}/vpc-id/"
+     */
+    vpc_path_mem = flb_sds_create_size(70);
+    if (!vpc_path_mem) {
+        flb_errno();
+        flb_sds_destroy(mac_id);
         return -1;
     }
+    vpc_path = flb_sds_printf(&vpc_path_mem, "%s/%s/%s/",
+                              "/latest/meta-data/network/interfaces/macs",
+                              mac_id, "vpc-id");
+    if (!vpc_path) {
+        flb_errno();
+        flb_sds_destroy(mac_id);
+        flb_sds_destroy(vpc_path_mem);
+        return -1;
+    }
+    ret = flb_aws_imds_request(ctx->client_imds, vpc_path, &ctx->vpc_id, &ctx->vpc_id_len);
 
-    flb_http_add_header(client, FLB_FILTER_AWS_IMDS_V2_TOKEN_TTL_HEADER,
-                        FLB_FILTER_AWS_IMDS_V2_TOKEN_TTL_HEADER_LEN,
-                        FLB_FILTER_AWS_IMDS_V2_TOKEN_TTL_HEADER_VAL,
-                        FLB_FILTER_AWS_IMDS_V2_TOKEN_TTL_HEADER_VAL_LEN);
+    flb_sds_destroy(mac_id);
+    flb_sds_destroy(vpc_path_mem);
 
-    /* Perform request */
-    ret = flb_http_do(client, &b_sent);
-    flb_plg_debug(ctx->ins, "IMDSv2 token request http_do=%i, HTTP Status: %i",
-              ret, client->resp.status);
+    return ret;
+}
 
-    if (ret != 0 || client->resp.status != 200) {
-        if (client->resp.payload_size > 0) {
-            flb_plg_debug(ctx->ins, "IMDSv2 token response\n%s",
-                          client->resp.payload);
+void flb_filter_aws_tags_destroy(struct flb_filter_aws *ctx)
+{
+    size_t i;
```

Comment:
why size_t and not int for loop iterator?

Reply:
No reason, changed to `int`, https://github.com/fluent/fluent-bit/pull/6065/commits/f6913d3bccf2e4fac1bcf3ab2505133f3fba7f34